app/src/org/commcare/dalvik/activities/EntitySelectActivity.java

```diff
@@ -1233,11 +1241,48 @@ public static HereFunctionHandler getHereFunctionHandler() {
         return hereFunctionHandler;
     }
 
-    public boolean getContainsHereFunction() {
-        return containsHereFunction;
+    public void onHereFunctionEvaluated() {
+        if (!containsHereFunction) {  // First time here() is evaluated
+            hereFunctionHandler.refreshLocation();
+            hereFunctionHandler.allowGpsUse();
+            containsHereFunction = true;
+
+            if (!hereFunctionHandler.locationProvidersFound()) {
+                locationNotificationHandler.sendEmptyMessage(0);
+            }
+        }
     }
 
-    public void setContainsHereFunction(boolean containsHereFunction) {
-        this.containsHereFunction = containsHereFunction;
+    private static class LocationNotificationHandler extends Handler {
+        private final WeakReference<EntitySelectActivity> mActivity;
+
+        public LocationNotificationHandler(EntitySelectActivity activity) {
+            mActivity = new WeakReference<>(activity);
+        }
+
+        @Override
+        public void handleMessage(Message msg) {
+            super.handleMessage(msg);
+
+            final EntitySelectActivity activity = mActivity.get();
+            if (activity != null) {
+                DialogInterface.OnClickListener onChangeListener = new DialogInterface.OnClickListener() {
+                    public void onClick(DialogInterface dialog, int i) {
+                        switch (i) {
+                            case DialogInterface.BUTTON_POSITIVE:
+                                Intent intent = new Intent(android.provider.Settings.ACTION_LOCATION_SOURCE_SETTINGS);
+                                activity.startActivity(intent);
+                                hereFunctionHandler.allowGpsUse();

```

Comment:
This gets fired before the user has the chance to actually turn on location services. That might be fine, or you could do `activity.startActivityForResult`; not sure the addtional complexity is worth it though